pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ReaderHandler.java

```diff
@@ -0,0 +1,260 @@
+
+            int pending = pendingMessages.incrementAndGet();
+            if (pending < maxPendingMessages) {
+                // Start next read in a separate thread to avoid recursion
+                service.getExecutor().execute(() -> receiveMessage());
+            }
+        }).exceptionally(exception -> {
+            return null;
+        });
+    }
+
+    @Override
+    public void onWebSocketText(String message) {
+        super.onWebSocketText(message);
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (reader != null) {
+            this.service.removeReader(this);
+            reader.closeAsync().thenAccept(x -> {
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}] Closed reader asynchronously", reader.getTopic());
+                }
+            }).exceptionally(exception -> {
+                log.warn("[{}] Failed to close reader", reader.getTopic(), exception);
+                return null;
+            });
+        }
+    }
+    
+    public Consumer getConsumer() {
+        return ((ReaderImpl)reader).getConsumer();
+    }
+
+    public String getSubscription() {
+        return subscription;
+    }
+    
+    public SubscriptionType getSubscriptionType() {
+        return SubscriptionType.Exclusive;
+    }
+
+    public long getAndResetNumMsgsDelivered() {
+        return numMsgsDelivered.sumThenReset();
+    }
+
+    public long getAndResetNumBytesDelivered() {
+        return numBytesDelivered.sumThenReset();
+    }
+
+    public long getMsgDeliveredCounter() {
+        return MSG_DELIVERED_COUNTER_UPDATER.get(this);
+    }
+    
+    protected void updateDeliverMsgStat(long msgSize) {
+        numMsgsDelivered.increment();
+        MSG_DELIVERED_COUNTER_UPDATER.incrementAndGet(this);
+        numBytesDelivered.add(msgSize);
+    }
+
+    private ReaderConfiguration getReaderConfiguration() {
+        ReaderConfiguration conf = new ReaderConfiguration();
+
+        if (queryParams.containsKey("readerName")) {
+            conf.setReaderName(queryParams.get("readerName"));
+        }
+
+        if (queryParams.containsKey("receiverQueueSize")) {
+            conf.setReceiverQueueSize(Math.min(Integer.parseInt(queryParams.get("receiverQueueSize")), 1000));
+        }
+        return conf;
+    }
+
+    @Override
+    protected Boolean isAuthorized(String authRole) throws Exception {
+        return service.getAuthorizationManager().canConsume(DestinationName.get(topic), authRole);
+    }
+
+    private static String checkRequestURI(HttpServletRequest request) {
+        String uri = request.getRequestURI();
+        List<String> parts = Splitter.on("/").splitToList(uri);
+
+        // Format must be like :
+        // /ws/reader/persistent/my-property/my-cluster/my-ns/my-topic
+        checkArgument(parts.size() == 8, "Invalid topic name format");
+        checkArgument(parts.get(1).equals("ws"));
+        checkArgument(parts.get(3).equals("persistent"));
```

Comment:
What about `non-persistent` topics? I think we should support them as well in the WebSocket interface